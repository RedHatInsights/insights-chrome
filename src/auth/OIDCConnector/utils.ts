import { AuthContextProps } from 'react-oidc-context';
import { ITLess, LOGIN_SCOPES_STORAGE_KEY, deleteLocalStorageItems } from '../../utils/common';
import { GLOBAL_FILTER_KEY, OFFLINE_REDIRECT_STORAGE_KEY } from '../../utils/consts';
import Cookies from 'js-cookie';
import logger from '../logger';
import createUUID from './createUUID';

const log = logger('auth:utils');

enum AllowedPartnerScopes {
  aws = 'aws',
  azure = 'azure',
  gcp = 'gcp',
}

function isPartnerScope(scope: string): scope is AllowedPartnerScopes {
  return Object.values(AllowedPartnerScopes).includes(scope as AllowedPartnerScopes);
}

function getPartnerScope(pathname: string) {
  // replace beta and leading "/"
  const sanitizedPathname = pathname.replace(/^(\/beta\/|\/preview\/)/, '/').replace(/^\//, '');
  // check if the pathname is connect/:partner
  if (sanitizedPathname.match(/^connect\/.+/)) {
    // return :partner param
    const fragmentScope = sanitizedPathname.split('/')[1];
    if (isPartnerScope(fragmentScope)) {
      return `api.partner_link.${fragmentScope}`;
    }
    log(`Invalid stratosphere scope: ${fragmentScope}`);
    return undefined;
  }

  return undefined;
}

export async function logout(auth: AuthContextProps, bounce?: boolean) {
  const keys = Object.keys(localStorage).filter(
    (key) =>
      key.endsWith('/api/entitlements/v1/services') ||
      key.endsWith('/chrome') ||
      key.endsWith('/chrome-store') ||
      key.startsWith('kc-callback') ||
      key.startsWith(GLOBAL_FILTER_KEY)
  );
  deleteLocalStorageItems([...keys, OFFLINE_REDIRECT_STORAGE_KEY, LOGIN_SCOPES_STORAGE_KEY]);
  if (bounce) {
    const eightSeconds = new Date(new Date().getTime() + 8 * 1000);
    Cookies.set('cs_loggedOut', 'true', {
      expires: eightSeconds,
    });
    await auth.signoutRedirect({
      redirectTarget: 'top',
      post_logout_redirect_uri: window.location.origin,
      id_token_hint: undefined,
    });
  } else {
    await auth.revokeTokens(['access_token', 'refresh_token']);
  }
}

export function login(auth: AuthContextProps, requiredScopes: string[] = [], redirectUri = location.href) {
  log('Logging in');
  // Redirect to login
  Cookies.set('cs_loggedOut', 'false');
  //FIX ME: Temp fix until scope is added in-boundary
  let scope = ITLess() ? ['openid', ...requiredScopes] : ['openid', ...requiredScopes];
  const partner = getPartnerScope(window.location.pathname);
  if (partner) {
    scope.push(partner);
  }
  scope = Array.from(new Set(scope));
  localStorage.setItem(LOGIN_SCOPES_STORAGE_KEY, JSON.stringify(scope));
  // KC scopes are delimited by a space character, hence the join(' ')
  return auth.signinRedirect({
    redirect_uri: redirectUri,
    scope: scope.join(' '),
    nonce: createUUID(),
  });
}
