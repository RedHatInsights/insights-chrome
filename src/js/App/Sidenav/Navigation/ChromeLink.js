import React, { useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';

import { appNavClick } from '../../../redux/actions';
import NavContext from './navContext';

const useDynamicModule = (appId) => {
  const [isDynamic, setIsDynamic] = useState();
  const { modules, activeModule } = useSelector(({ chrome: { modules = {}, activeModule } }) => ({
    modules,
    activeModule,
  }));
  useEffect(() => {
    const currentModule = modules[appId];
    if (!currentModule) {
      setIsDynamic(false);
    } else if (appId === activeModule) {
      setIsDynamic(true);
    } else if (currentModule && appId !== activeModule) {
      setIsDynamic(currentModule.dynamic !== false && modules[activeModule]?.dynamic !== false);
    }
  }, [appId]);

  return isDynamic;
};

const LinkWrapper = ({ href, isBeta, onLinkClick, className, currAppId, appId, children }) => {
  let actionId = href.split('/').slice(2).join('/');
  if (actionId.includes('/')) {
    actionId = actionId.split('/').pop();
  }
  if (currAppId !== appId && href.split('/').length === 3) {
    actionId = '/';
  }

  /**
   * If the sub nav item points to application root
   * eg. /openshift/cost-management we don't want to send "/cost-management" but "/"
   * We are not in app sub route but in app root
   */
  const domEvent = {
    href,
    id: actionId,
    navId: actionId,
    /**
     * @deprecated
     * Remove once nav overhaul is in all environments
     */
    type: 'click',
  };
  const dispatch = useDispatch();
  const onClick = (event) => {
    if (onLinkClick && isBeta) {
      if (!onLinkClick(event, href)) {
        return false;
      }
    }

    dispatch(appNavClick({ id: actionId }, domEvent));
  };
  return (
    <NavLink data-testid="router-link" onClick={onClick} to={href} className={className}>
      {children}
    </NavLink>
  );
};

LinkWrapper.propTypes = {
  href: PropTypes.string.isRequired,
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
  isBeta: PropTypes.bool,
  onLinkClick: PropTypes.func.isRequired,
  currAppId: PropTypes.string,
  appId: PropTypes.string.isRequired,
};

const basepath = document.baseURI;

const cleanRefreshLinkProps = ({ active, onClick, appId, currAppId, ...rest }) => rest;

const RefreshLink = (props) => {
  const { href, isExternal, onLinkClick, isBeta, ...rest } = cleanRefreshLinkProps(props);
  return (
    <a
      data-testid="native-link"
      href={isExternal ? href : `${basepath}${href.replace(/^\//, '')}`}
      {...(isExternal
        ? {
            rel: 'noreferrer noopener',
            target: '_blank',
          }
        : {})}
      onClick={(event) => {
        if (onLinkClick && isBeta && !isExternal) {
          if (!onLinkClick(event, href)) {
            return false;
          }
        }
      }}
      {...rest}
    />
  );
};

RefreshLink.propTypes = {
  href: PropTypes.string.isRequired,
  isExternal: PropTypes.bool,
  onLinkClick: PropTypes.func,
  isBeta: PropTypes.bool,
};

const ChromeLink = ({ appId, children, ...rest }) => {
  const { onLinkClick } = useContext(NavContext);
  const currAppId = useSelector(({ chrome }) => chrome?.appId);
  const isDynamic = useDynamicModule(appId);

  if (!rest.isExternal && typeof isDynamic === 'undefined') {
    return null;
  }

  const LinkComponent = !rest.isExternal && isDynamic ? LinkWrapper : RefreshLink;
  return (
    <LinkComponent onLinkClick={onLinkClick} appId={appId} currAppId={currAppId} {...rest}>
      {children}
    </LinkComponent>
  );
};

ChromeLink.propTypes = {
  appId: PropTypes.string,
  children: PropTypes.node.isRequired,
};

export default ChromeLink;
