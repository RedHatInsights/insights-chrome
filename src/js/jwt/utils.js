const pub = {};

pub.exposeTest = (priv) => {
    if (window.NODE_ENV === 'test') {
        module.exports.priv = priv;
    }
};

module.exports = pub;
