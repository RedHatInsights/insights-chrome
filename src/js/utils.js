import get from 'lodash/get';

function getWindow() {
    return window;
}

/* eslint-disable curly */
export function isValidAccountNumber(num) {
    if (!num) return false;
    if (num === -1) return false;
    if (num === '-1') return false;
    return Number.isInteger(Number(num));
}
/* eslint-enable curly */

function getSection() {
    const sections = getWindow().location.pathname.split('/');
    if (sections[1] === 'beta') { return sections[2] || ''; }

    return sections[1];
}

export function pageAllowsUnentitled() {
    const pathname = getWindow().location.pathname;
    if (pathname === '/' ||
        pathname === '/beta' ||
        pathname === '/beta/' ||
        pathname.indexOf('/openshift') === 0 ||
        pathname.indexOf('/beta/openshift') === 0) {
        return true;
    }

    return false;
}

export function pageRequiresAuthentication() {
    const section = getSection();
    if (section === 'insights' ||
        section === 'rhel'     ||
        section === 'hybrid'   ||
        section === 'apps'     ||
        section === 'ansible'  ||
        section === 'migrations') {
        return true;
    }

    return false;
}

/**
 * Creates a redux listener that watches the state on given path (e.g. chrome.appNav) and calls
 * the given function when the state on the given path changes.
 *
 * The function is called with two parameters: current state value on the path, store reference
 */
export function createReduxListener(store, path, fn) {
    let previous = undefined;

    return () => {
        const state = store.getState();
        const current = get(state, path);

        if (current !== previous) {
            previous = current;
            fn(current, store);
        }
    };
}
