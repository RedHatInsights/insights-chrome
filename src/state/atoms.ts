import { atom } from 'jotai';
// setup initial chrome atoms
export const contextSwitcherOpenAtom = atom(false);
export const activeModuleAtom = atom<string | undefined>(undefined);
