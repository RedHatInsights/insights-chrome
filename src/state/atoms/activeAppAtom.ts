import { NavDOMEvent } from '@redhat-cloud-services/types';
import { atom } from 'jotai';

export type NavEvent = { navId?: string; domEvent: NavDOMEvent };
export type NavListener = (navEvent: NavEvent) => void;

export const activeAppAtom = atom<string | undefined>(undefined);
export const activeNavListenersAtom = atom<{ [listenerId: number]: NavListener | undefined }>({});
export const addNavListenerAtom = atom(null, (_get, set, navListener: NavListener) => {
  const listenerId = Date.now();
  set(activeNavListenersAtom, (prev) => {
    return { ...prev, [listenerId]: navListener };
  });
  return listenerId;
});
export const deleteNavListenerAtom = atom(null, (get, set, id: number) => {
  set(activeNavListenersAtom, (prev) => {
    return { ...prev, [id]: undefined };
  });
});

export const triggerNavListenersAtom = atom(null, (get, _set, event: NavEvent) => {
  const activeNavListeners = get(activeNavListenersAtom);
  Object.values(activeNavListeners).forEach((el) => {
    el?.(event);
  });
});
