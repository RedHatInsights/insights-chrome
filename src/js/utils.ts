import get from 'lodash/get';
import { Store } from 'redux';
import { DEFAULT_ROUTES } from './jwt/constants';
import flatMap from 'lodash/flatMap';
import { NavItem } from './types';
import axios from 'axios';
import { ChromeModule, RouteDefinition } from './redux/store';

export function getWindow() {
  return window;
}

export function isValidAccountNumber(num?: number | string) {
  if (!num) return false;
  if (num === -1) return false;
  if (num === '-1') return false;
  return Number.isInteger(Number(num));
}

export function getSection() {
  const sections = getWindow().location.pathname.split('/');
  if (sections[1] === 'beta') {
    return sections[2] || '';
  }

  return sections[1];
}

export function pageAllowsUnentitled() {
  const pathname = getWindow().location.pathname;
  if (
    pathname === '/' ||
    pathname === '/beta' ||
    pathname === '/beta/' ||
    pathname.indexOf('/openshift') === 0 ||
    pathname.indexOf('/beta/openshift') === 0 ||
    pathname.indexOf('/security') === 0 ||
    pathname.indexOf('/beta/security') === 0 ||
    pathname.indexOf('/application-services') === 0 ||
    pathname.indexOf('/beta/application-services') === 0 ||
    pathname.indexOf('/hac') === 0 ||
    pathname.indexOf('/beta/hac') === 0 ||
    pathname.indexOf('/ansible/ansible-dashboard/trial') === 0 ||
    pathname.indexOf('/beta/ansible/ansible-dashboard/trial') === 0
  ) {
    return true;
  }

  return false;
}

export function pageRequiresAuthentication() {
  const section = getSection();
  if (
    section === 'insights' ||
    section === 'cost-management' ||
    section === 'apps' ||
    section === 'ansible' ||
    section === 'migrations' ||
    section === 'subscriptions' ||
    section === 'openshift' ||
    section === 'settings' ||
    section === 'user-preferences' ||
    section === 'internal' ||
    section === 'application-services'
  ) {
    return true;
  }

  return false;
}

/**
 * Creates a redux listener that watches the state on given path (e.g. chrome.appNav) and calls
 * the given function when the state on the given path changes.
 *
 * The function is called with two parameters: current state value on the path, store reference
 */
export function createReduxListener(store: Store, path: string, fn: (current: any, store: Store) => void) {
  let previous: any = undefined;

  return () => {
    const state = store.getState();
    const current = get(state, path);

    if (current !== previous) {
      previous = current;
      fn(current, store);
    }
  };
}

export function deleteLocalStorageItems(keys: string[]) {
  keys.map((key) => localStorage.removeItem(key));
}

export function lastActive(searchString: string, fallback: string) {
  return Object.keys(localStorage).reduce<string | { expires: string }>((acc, curr) => {
    if (curr.includes(searchString)) {
      try {
        let accDate;
        try {
          const localStorageDate = localStorage.getItem(acc as string);
          accDate = localStorageDate ? new Date(JSON.parse(localStorageDate)?.expires) : new Date();
        } catch {
          accDate = new Date();
        }

        const currObj = JSON.parse(localStorage.getItem(curr) || '');
        return accDate >= new Date(currObj.expires) ? acc : curr;
      } catch (e) {
        return acc;
      }
    }

    return acc;
  }, fallback);
}

export const isAnsible = (sections: string[]) => (sections.includes('ansible') && sections.includes('insights') ? 1 : 0);

export function getUrl(type?: string) {
  if (window.location.pathname === '/beta/' || window.location.pathname === '/') {
    return 'landing';
  }

  const sections = window.location.pathname.split('/');
  if (sections[1] === 'beta') {
    return type === 'bundle' ? sections[2] : sections[3 + isAnsible(sections)];
  }

  return type === 'bundle' ? sections[1] : sections[2 + isAnsible(sections)];
}

export function getEnv() {
  return Object.entries(DEFAULT_ROUTES).find(([, { url }]) => url.includes(location.hostname))?.[0] || 'qa';
}

export function getEnvDetails() {
  return Object.entries(DEFAULT_ROUTES).find(([, { url }]) => url.includes(location.hostname))?.[1];
}

export function isProd() {
  return location.host === 'cloud.redhat.com' || location.host === 'console.redhat.com' || location.host.includes('prod.foo.redhat.com');
}

export function isBeta() {
  return window.location.pathname.split('/')[1] === 'beta' ? true : false;
}

export function isFedRamp() {
  return getEnv() === 'gov';
}

export function updateDocumentTitle(title?: string, noSuffix = false) {
  const titleSuffix = '| console.redhat.com';
  if (typeof title === 'undefined') {
    return;
  }
  if (typeof title === 'string') {
    document.title = title.includes(titleSuffix) || noSuffix ? title : `${title} ${titleSuffix}`;
  } else {
    console.warn(`Title is not a string. Got ${typeof title} instead.`);
  }
}

const activateChild = (hrefMatch: string, childRoutes: NavItem[]) => {
  let hasActiveChild = false;
  const routes = childRoutes.map((item) => {
    const active = item.href === hrefMatch;
    if (active) {
      hasActiveChild = true;
    }
    return {
      ...item,
      active,
    };
  });
  return {
    active: hasActiveChild,
    routes,
  };
};

function mutateSchema(hrefMatch: string, navItems: NavItem[]): NavItem[] {
  return navItems.map((item) => {
    const { href, routes, navItems } = item;
    if (!href && navItems) {
      return {
        ...item,
        navItems: mutateSchema(hrefMatch, navItems),
      };
    }

    if (!href && routes) {
      return {
        ...item,
        ...activateChild(hrefMatch, routes),
      };
    }

    if (href) {
      return {
        ...item,
        active: item.href === hrefMatch,
      };
    }

    return item;
  });
}

export const highlightItems = (pathname: string, navItems: NavItem[], sortedLinks: string[]) => {
  const cleanPathname = pathname.replace(/\/$/, '');
  const segmentsCount = cleanPathname.split('/').length + 1;
  const matchedLink = sortedLinks.find((href) => {
    const segmentedHref = href.replace(/\/$/, '').split('/').slice(0, segmentsCount).join('/');
    return cleanPathname.includes(segmentedHref);
  });
  return mutateSchema(matchedLink?.replace(/\/$/, '') || '', navItems);
};

export const levelArray = (navItems: NavItem[]): string[] => {
  return flatMap<NavItem, string>(navItems, ({ href, routes, navItems }) => {
    if (!href && navItems) {
      return levelArray(navItems);
    }

    if (!href && routes) {
      return levelArray(routes);
    }

    if (href) {
      return [href];
    }

    return [];
  });
};

// eslint-disable-next-line @typescript-eslint/no-empty-function
export function noop() {}

export const trustarcScriptSetup = () => {
  const trustarcScript = document.createElement('script');
  trustarcScript.id = 'trustarc';

  if (location.host === 'console.redhat.com') {
    trustarcScript.src = '//static.redhat.com/libs/redhat/marketing/latest/trustarc/trustarc.js';
  } else {
    trustarcScript.src = '//static.redhat.com/libs/redhat/marketing/latest/trustarc/trustarc.stage.js';
  }

  document.body.appendChild(trustarcScript);
};

export const loadFedModules = () =>
  axios.get(`${window.location.origin}${isBeta() ? '/beta' : ''}/config/chrome/fed-modules.json?ts=${Date.now()}`, {
    headers: {
      'Cache-Control': 'no-cache',
      Pragma: 'no-cache',
      Expires: '0',
    },
  });

export const generateRoutesList = (modules: { [key: string]: ChromeModule }) =>
  Object.entries(modules)
    .reduce<RouteDefinition[]>(
      (acc, [scope, { dynamic, manifestLocation, isFedramp, modules = [] }]) => [
        ...acc,
        ...modules
          .map(({ module, routes }) =>
            /**Clean up this map function */
            routes.map((route) => ({
              scope,
              module,
              isFedramp: typeof route === 'string' ? isFedramp : route.isFedramp,
              path: typeof route === 'string' ? route : route.pathname,
              manifestLocation,
              dynamic: typeof dynamic === 'boolean' ? dynamic : typeof route === 'string' ? true : route.dynamic,
              exact: typeof route === 'string' ? false : route.exact,
            }))
          )
          .flat(),
      ],
      []
    )
    .sort((a, b) => (a.path.length < b.path.length ? 1 : -1));
