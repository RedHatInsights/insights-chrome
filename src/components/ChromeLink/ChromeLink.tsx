import React, { memo, useContext, useMemo, useRef } from 'react';
import { NavLink } from 'react-router-dom';
import { preloadModule } from '@scalprum/core';

import NavContext, { OnLinkClick } from '../Navigation/navContext';
import { NavDOMEvent } from '../../@types/types';
import { useAtomValue, useSetAtom } from 'jotai';
import { activeModuleAtom } from '../../state/atoms/activeModuleAtom';
import { moduleRoutesAtom } from '../../state/atoms/chromeModuleAtom';
import { triggerNavListenersAtom } from '../../state/atoms/activeAppAtom';

interface RefreshLinkProps extends React.HTMLAttributes<HTMLAnchorElement> {
  isExternal?: boolean;
  onLinkClick?: OnLinkClick;
  isBeta?: boolean;
  href: string;
  active?: boolean;
  onClick?: () => void;
  appId?: string;
  currAppId?: string;
  target?: string;
  rel?: string;
}

export interface LinkWrapperProps extends RefreshLinkProps {
  className?: string;
  tabIndex?: number;
}

const LinkWrapper: React.FC<LinkWrapperProps> = memo(
  ({ href = '', isBeta, onLinkClick, className, currAppId, appId, children, tabIndex, ...props }) => {
    const linkRef = useRef<HTMLAnchorElement | null>(null);
    const moduleRoutes = useAtomValue(moduleRoutesAtom);
    const triggerNavListener = useSetAtom(triggerNavListenersAtom);
    const moduleEntry = useMemo(() => moduleRoutes?.find((route) => href?.includes(route.path)), [href, appId]);
    const preloadTimeout = useRef<NodeJS.Timeout>();
    let actionId = href.split('/').slice(2).join('/');
    if (actionId.includes('/')) {
      actionId = actionId.split('/').pop() as string;
    }
    if (currAppId !== appId && href.split('/').length === 3) {
      actionId = '/';
    }

    /**
     * If the sub nav item points to application root
     * eg. /openshift/cost-management we don't want to send "/cost-management" but "/"
     * We are not in app sub route but in app root
     */
    const domEvent: NavDOMEvent = {
      href,
      id: actionId,
      navId: actionId,
      /**
       * @deprecated
       * Remove once nav overhaul is in all environments
       */
      type: 'click',
    };
    const onClick = (event: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => {
      if (event.ctrlKey || event.shiftKey) {
        return false;
      }
      if (onLinkClick && isBeta) {
        if (!onLinkClick(event, href)) {
          return false;
        }
      }

      /**
       * Add reference to the DOM link element
       */
      domEvent.target = linkRef.current;
      triggerNavListener({ navId: actionId, domEvent });
    };

    // turns /settings/rbac/roles -> settings_rbac_roles
    const quickStartHighlightId = href
      .split('/')
      .slice(href.startsWith('/') ? 1 : 0)
      .join('_');
    return (
      <NavLink
        {...props}
        onMouseEnter={() => {
          if (moduleEntry) {
            preloadTimeout.current = setTimeout(() => {
              preloadModule(moduleEntry?.scope, moduleEntry?.module);
            }, 250);
          }
        }}
        onMouseLeave={() => {
          if (preloadTimeout.current) {
            clearTimeout(preloadTimeout.current);
          }
        }}
        tabIndex={tabIndex}
        ref={linkRef}
        data-testid="router-link"
        onClick={onClick}
        to={href}
        className={className}
        data-quickstart-id={quickStartHighlightId}
      >
        {children}
      </NavLink>
    );
  }
);

LinkWrapper.displayName = 'MemoizedLinkWrapper';

const basepath = document.baseURI;

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const cleanRefreshLinkProps = ({ active, onClick, appId, currAppId, ...rest }: RefreshLinkProps) => rest;

const RefreshLink: React.FC<RefreshLinkProps> = (props) => {
  const { href, isExternal, onLinkClick, isBeta, ...rest } = cleanRefreshLinkProps(props);
  return (
    <a
      data-testid="native-link"
      href={isExternal ? href : `${basepath}${href.replace(/^\//, '')}`}
      {...(isExternal
        ? {
            rel: 'noreferrer noopener',
            target: '_blank',
          }
        : {})}
      onClick={(event) => {
        if (onLinkClick && isBeta && !isExternal) {
          if (!onLinkClick(event, href)) {
            return false;
          }
        }
      }}
      {...rest}
    />
  );
};

const ChromeLink: React.FC<LinkWrapperProps> = ({ appId, children, ...rest }) => {
  const { onLinkClick, isNavOpen, inPageLayout } = useContext(NavContext);
  const currAppId = useAtomValue(activeModuleAtom);

  const LinkComponent = !rest.isExternal ? LinkWrapper : RefreshLink;
  return (
    <LinkComponent {...(inPageLayout && !isNavOpen ? { tabIndex: -1 } : {})} onLinkClick={onLinkClick} appId={appId} currAppId={currAppId} {...rest}>
      {children}
    </LinkComponent>
  );
};

export default ChromeLink;
