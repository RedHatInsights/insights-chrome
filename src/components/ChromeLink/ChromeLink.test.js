/* eslint-disable react/prop-types */
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import { Provider as JotaiProvider } from 'jotai';
import { useHydrateAtoms } from 'jotai/utils';

import ChromeLink from './ChromeLink';
import NavContext from '../Navigation/navContext';
import { activeModuleAtom } from '../../state/atoms/activeModuleAtom';
import { moduleRoutesAtom } from '../../state/atoms/chromeModuleAtom';
import { triggerNavListenersAtom } from '../../state/atoms/activeAppAtom';

const HydrateAtoms = ({ initialValues, children }) => {
  useHydrateAtoms(initialValues);
  return children;
};

const Wrapper = ({ children, atomValues = [] }) => (
  <MemoryRouter>
    <JotaiProvider>
      <HydrateAtoms initialValues={atomValues}>
        <NavContext.Provider value={{ onLinkClick: jest.fn() }}>{children}</NavContext.Provider>
      </HydrateAtoms>
    </JotaiProvider>
  </MemoryRouter>
);

describe('ChromeLink', () => {
  const defaultAtomValues = [
    [activeModuleAtom, 'testModule'],
    [moduleRoutesAtom, []],
    [triggerNavListenersAtom, jest.fn()],
  ];

  test('should render chrome link', () => {
    const { container } = render(
      <Wrapper atomValues={defaultAtomValues}>
        <ChromeLink href="/test">Test Link</ChromeLink>
      </Wrapper>
    );
    expect(container).toMatchSnapshot();
  });

  test('should render external chrome link', () => {
    const { container } = render(
      <Wrapper atomValues={defaultAtomValues}>
        <ChromeLink href="https://external.com" isExternal>
          External Link
        </ChromeLink>
      </Wrapper>
    );
    expect(container).toMatchSnapshot();
  });
});
