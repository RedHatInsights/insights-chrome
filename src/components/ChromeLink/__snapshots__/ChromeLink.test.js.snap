// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChromeLink should render chrome link 1`] = `
<div>
  <a
    class=""
    data-quickstart-id="test"
    data-testid="router-link"
    href="/test"
  >
    Test Link
  </a>
</div>
`;

exports[`ChromeLink should render external chrome link 1`] = `
<div>
  <a
    data-testid="native-link"
    href="https://external.com"
    rel="noreferrer noopener"
    target="_blank"
  >
    External Link
  </a>
</div>
`;
